src/main/target/SITL/target.c

```diff
@@ -0,0 +1,411 @@
+}
+void delayMicroseconds(uint32_t us) {
+	microsleep(us / simRate);
+}
+void delayMicroseconds_real(uint32_t us) {
+	microsleep(us);
+}
+void delay(uint32_t ms) {
+	uint64_t start = millis64();
+
+	while ((millis64() - start) < ms) {
+		microsleep(1000);
+	}
+}
+
+
+// PWM part
+bool pwmMotorsEnabled = false;
+static pwmOutputPort_t motors[MAX_SUPPORTED_MOTORS];
+static pwmOutputPort_t servos[MAX_SUPPORTED_SERVOS];
+
+// real value to send
+static uint16_t motorsPwm[MAX_SUPPORTED_MOTORS];
+static uint16_t servosPwm[MAX_SUPPORTED_SERVOS];
+static uint16_t idlePulse;
+
+void motorDevInit(const motorDevConfig_t *motorConfig, uint16_t _idlePulse, uint8_t motorCount) {
+	UNUSED(motorConfig);
+	UNUSED(motorCount);
+
+	idlePulse = _idlePulse;
+
+	for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+		motors[motorIndex].enabled = true;
+	}
+	pwmMotorsEnabled = true;
+}
+void servoDevInit(const servoDevConfig_t *servoConfig) {
+	UNUSED(servoConfig);
+	for (uint8_t servoIndex = 0; servoIndex < MAX_SUPPORTED_SERVOS; servoIndex++) {
+		servos[servoIndex].enabled = true;
+	}
+}
+
+pwmOutputPort_t *pwmGetMotors(void) {
+	return motors;
+}
+bool pwmAreMotorsEnabled(void) {
+	return pwmMotorsEnabled;
+}
+void pwmWriteMotor(uint8_t index, uint16_t value) {
+	motorsPwm[index] = value - idlePulse;
+}
+void pwmShutdownPulsesForAllMotors(uint8_t motorCount) {
+	UNUSED(motorCount);
+	pwmMotorsEnabled = false;
+}
+void pwmCompleteMotorUpdate(uint8_t motorCount) {
+	UNUSED(motorCount);
+	// send to simulator
+	// for gazebo8 ArduCopterPlugin remap, range = [0.0, 1.0]
+	pwmPkt.motor_speed[3] = motorsPwm[0] / 1000.0f;
+	pwmPkt.motor_speed[0] = motorsPwm[1] / 1000.0f;
+	pwmPkt.motor_speed[1] = motorsPwm[2] / 1000.0f;
+	pwmPkt.motor_speed[2] = motorsPwm[3] / 1000.0f;
+
+	// get one "fdm_packet" can only send one "servo_packet"!!
+	if(pthread_mutex_trylock(&updateLock) != 0) return;
+	udpSend(&pwmLink, &pwmPkt, sizeof(servo_packet));
+//	printf("[pwm]%u:%u,%u,%u,%u\n", idlePulse, motorsPwm[0], motorsPwm[1], motorsPwm[2], motorsPwm[3]);
+}
+void pwmWriteServo(uint8_t index, uint16_t value) {
+	servosPwm[index] = value;
+}
+
+// ADC part
+uint16_t adcGetChannel(uint8_t channel) {
+	UNUSED(channel);
+	return 0;
+}
+
+// stack part
+char _estack;
+char _Min_Stack_Size;
+
+// fake EEPROM
+extern uint8_t __config_start;
+extern uint32_t __FLASH_CONFIG_Size;
+static FILE *eepromFd = NULL;
+const char *EEPROM_FILE = EEPROM_FILENAME;
+
+void FLASH_Unlock(void) {
+	uint8_t * const eeprom = &__config_start;
+
+	if(eepromFd != NULL) {
+		printf("[FLASH_Unlock] eepromFd != NULL\n");
+		return;
+	}
+
+	// open or create
+	eepromFd = fopen(EEPROM_FILE,"r+");
+	if(eepromFd != NULL) {
+		long lSize;
+		int c;
+
+		// obtain file size:
+		fseek(eepromFd , 0 , SEEK_END);
+		lSize = ftell(eepromFd);
+		rewind(eepromFd);
+
+		printf("[FLASH_Unlock]size = %ld\n", lSize);
+		for(unsigned int i=0; i<((uintptr_t)&__FLASH_CONFIG_Size); i++){
```

Comment:
is lSize correct? The loop uses `__FLASH_CONFIG_Size`. 
(__config_end - __config_start) may be better option, __FLASH_CONFIG_Size is quite ugly (using symbol address as integer)

Another problem may be that eeprom is placed incorrectly in linker (INSERT AFTER .bss) and libc/gcc has some guard there

This is strange ...